azure_monitor/src/azure_monitor/sdk/auto_collection/live_metrics/manager.py

```diff
@@ -0,0 +1,170 @@
+
+from azure_monitor.sdk.auto_collection import live_metrics
+from azure_monitor.sdk.auto_collection.live_metrics.exporter import (
+    LiveMetricsExporter,
+)
+from azure_monitor.sdk.auto_collection.live_metrics.sender import (
+    LiveMetricsSender,
+)
+
+FALLBACK_INTERVAL = 60
+PING_INTERVAL = 5
+POST_INTERVAL = 1
+MAIN_INTERVAL = 2
+
+
+class LiveMetricsManager(threading.Thread):
+    """Live Metrics Manager
+
+    It will start Live Metrics process when instantiated,
+    responsible for switching between ping and post actions.
+    """
+
+    daemon = True
+
+    def __init__(self, meter, instrumentation_key):
+        super().__init__()
+        self.thread_event = threading.Event()
+        self.interval = MAIN_INTERVAL
+        self._instrumentation_key = instrumentation_key
+        self._is_user_subscribed = False
+        self._meter = meter
+        self._post = None
+        self._ping = LiveMetricsPing(self._instrumentation_key)
+        self.start()
+
+    def run(self):
+        self.check_if_user_is_subscribed()
+        while not self.thread_event.wait(self.interval):
+            self.check_if_user_is_subscribed()
+
+    def check_if_user_is_subscribed(self):
+        if self._ping:
+            if self._ping.is_user_subscribed:
+                # Switch to Post
+                self._ping.shutdown()
+                self._ping = None
+                self._post = LiveMetricsPost(
+                    self._meter, self._instrumentation_key
+                )
+        if self._post:
+            if not self._post.is_user_subscribed:
+                # Switch to Ping
+                self._post.shutdown()
+                self._post = None
+                self._ping = LiveMetricsPing(self._instrumentation_key)
+
+    def shutdown(self):
+        if self._ping:
+            self._ping.shutdown()
+        if self._post:
+            self._post.shutdown()
+        self.thread_event.set()
+
+
+class LiveMetricsPing(threading.Thread):
+    """Ping to Live Metrics service
+
+    Ping to determine if user is subscribed and live metrics need to be send.
+    """
+
+    daemon = True
+
+    def __init__(self, instrumentation_key):
+        super().__init__()
+        self.instrumentation_key = instrumentation_key
+        self.thread_event = threading.Event()
+        self.interval = PING_INTERVAL
+        self.is_user_subscribed = False
+        self.last_send_succeeded = False
+        self.last_request_success_time = 0
+        self.sender = LiveMetricsSender(self.instrumentation_key)
+        self.start()
+
+    def run(self):
+        self.ping()
+        while not self.thread_event.wait(self.interval):
+            self.ping()
+
+    def ping(self):
+        envelope = live_metrics.create_metric_envelope(
+            self.instrumentation_key
+        )
+        token = attach(set_value("suppress_instrumentation", True))
+        response = self.sender.ping(envelope)
+        detach(token)
+        if response.ok:
```

Comment:
What if no response from server?

Reply:
sender must handle timeout and failures and ok would be false, are you thinking about something else?

Reply:
Take a look at [this link](https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module). It is possible for `requests` to raise an exception.

Reply:
Ahh I see you are catching `HTTPError`. I think the other cases just need to be handled as well.